vite.plugins.ts

```diff
@@ -79,3 +81,60 @@ export const minifyHTML = (): {
     return minify(html, { collapseWhitespace: true });
   },
 });
+
+/**
+ * Forwards requests to the local replica.
+ * Denies access to raw URLs.
+ *
+ * @param replicaOrigin Replica URL to forward requests to
+ * @param forwardRules List of rules (i.e. hostname to canisterId mappings)
+ *                     to forward requests to a specific canister
+ */
+export const replicaForwardPlugin = (
+  replicaOrigin: string,
+  forwardRules: { hosts: string[]; canisterId: string }[]
+) => ({
+  name: "replica-forward",
+  configureServer(server: ViteDevServer) {
+    const proxy = httpProxy.createProxyServer({
+      secure: false,
+    });
+
+    server.middlewares.use((req, res, next) => {
+      if (
+        req.headers["host"]?.endsWith(".raw.ic0.app") ||
+        req.headers["host"]?.endsWith(".raw.icp0.io")
+      ) {
+        console.log(
+          `Denying access to raw URL ${req.method} https://${req.headers.host}${req.url}`
+        );
+        res.statusCode = 400;
+        res.end("Raw IC URLs are not supported");
+        return;
+      }
+
+      let matchingRule = forwardRules.find((rule) => {
+        let host = req.headers["host"];
+        return isNullish(host) || rule.hosts.includes(host);
+      });
+
+      if (matchingRule !== undefined) {
```

Comment:
for consistency you could use


```suggestion
      if (nonNullish(matchingRule)) {
```

and for even more consistency you could do something like this:

```ts
if(isNullish(matchingRule)) { 
  return next()
}
 
// happy/expected case
```